external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala

```diff
@@ -73,12 +73,19 @@ class DirectKafkaInputDStream[
 
   protected val kc = new KafkaCluster(kafkaParams)
 
-  protected val maxMessagesPerPartition: Option[Long] = {
-    val ratePerSec = context.sparkContext.getConf.getInt(
+  private val ratePerSec: Int = context.sparkContext.getConf.getInt(
       "spark.streaming.kafka.maxRatePerPartition", 0)
-    if (ratePerSec > 0) {
+  protected def maxMessagesPerPartition: Option[Long] = {
+    val estimatedRate = rateController.map(_.getLatestRate().toInt).getOrElse(-1)
+    val numPartitions = currentOffsets.keys.size
+    val effectiveRatePerSec = if (estimatedRate > 0) {
+      (ratePerSec min (estimatedRate / numPartitions))
```

Comment:
Ah, yes, that's "infix notation" and I also prefer `min(a,b)` for non-operators. I just pushed a minor fix to not use infix notation and also renamed `rarePerSec` to `ratePerPartition` for better consistency.